workflow/src/main/kotlin/com/r3/corda/sdk/token/workflow/flows/UpdateEvolvableToken.kt

```diff
@@ -1,40 +1,137 @@
+     */
+    @CordaSerializable
+    data class EvolvableTokenUpdateNotification(val signatureRequired: Boolean = false)
+
+    @InitiatingFlow
+    @StartableByRPC
+    class Initiator(
+            val oldStateAndRef: StateAndRef<EvolvableTokenType>,
+            val newState: EvolvableTokenType
+    ) : FlowLogic<SignedTransaction>() {
+
+        companion object {
+            object CREATING : ProgressTracker.Step("Creating update evolvable token transaction proposal.")
+
+            object SIGNING : ProgressTracker.Step("Signing transaction proposal.")
+
+            object COLLECTING : ProgressTracker.Step("Gathering counterparty signatures.") {
+                override fun childProgressTracker() = CollectSignaturesFlow.tracker()
+            }
+
+            object RECORDING : ProgressTracker.Step("Recording signed transaction.") {
+                override fun childProgressTracker() = FinalityFlow.tracker()
+            }
+
+            fun tracker() = ProgressTracker(CREATING, SIGNING, COLLECTING, RECORDING)
+        }
+
+        override val progressTracker: ProgressTracker = tracker()
+
+        @Suspendable
+        override fun call(): SignedTransaction {
+            // Create a transaction which updates the ledger with the new evolvable token.
+            // The parties listed as maintainers in the old state should be the signers.
+            // TODO Should this be both old and new maintainer lists?
+            progressTracker.currentStep = CREATING
+            val utx = subFlow(
+                    AddUpdateEvolvableToken(
+                            TransactionBuilder(notary = oldStateAndRef.state.notary),
+                            oldStateAndRef,
+                            newState
+                    )
+            )
+
+            // Sign the transaction proposal (creating a partially signed transaction, or ptx)
+            progressTracker.currentStep = SIGNING
+            val ptx: SignedTransaction = serviceHub.signInitialTransaction(utx)
+
+            // Gather signatures from other maintainers
+            progressTracker.currentStep = COLLECTING
+            val otherMaintainerSessions = otherMaintainers().map { initiateFlow(it) }
+            otherMaintainerSessions.forEach { it.send(EvolvableTokenUpdateNotification(signatureRequired = true)) }
+            val stx = subFlow(CollectSignaturesFlow(
+                    partiallySignedTx = ptx,
+                    sessionsToCollectFrom = otherMaintainerSessions,
+                    progressTracker = COLLECTING.childProgressTracker()
+            ))
+
+            // Distribute to all observers, including maintainers, participants, and subscribers (via distribution list)
+            progressTracker.currentStep = RECORDING
+            val observerSessions = wellKnownObservers().map { initiateFlow(it) }
+            observerSessions.forEach { it.send(EvolvableTokenUpdateNotification(signatureRequired = false)) }
```

Comment:
i wonder if there is a nicer way to handle this? Can't think of one off the top of my head.

Reply:
The logic is a bit convoluted due to the need to split maintainers from participants. I am very much open to better options!

Reply:
no worries was just a thought!